docs/architecture/2024-10-23_osquery_refactor.md

```diff
@@ -0,0 +1,79 @@
+
+In our current implementation, the extension is created by `runLauncher` and run via our rungroup, and passed into the osquery runner for use when creating the instances. To run multiple osquery instances, we would have to create multiple extensions inside `runLauncher`, pass them into the runner, and make sure we adequately map which extension belongs to which osquery instance in the process. This would also require a full launcher restart if the runner needed to increase the number of osquery instances running (for example, if a single-tenant launcher enrolls with a new tenant to become a multi-tenant launcher).
+
+## Decision
+
+To simplify the creation of multiple osquery instances, and to make it possible to increase the number of osquery instances running without performing a full launcher restart, we will move the osquery extension ownership to the osquery instance.
+
+This will also give us the opportunity to make the separation of responsibility between the runner and the instance clearer. The instance will create its own extension, extension server, etc. The runner will be responsible for the instances' state. This means the osquery instance stats should be owned by the runner, since the runner already has the responsibility of updating them.
+
+```mermaid
+---
+title: Refactored (one osquery instance)
+---
+erDiagram
+    RUNGROUP ||--|| OSQUERYRUNNER : runs
+    OSQUERYRUNNER ||--|| OSQUERYINSTANCE: runs
+    OSQUERYINSTANCE ||--|| EXTENSION : owns
+    EXTENSION ||--|| JSONRPCCLIENT : uses
+    OSQUERYINSTANCE ||--|{ EXTENSIONSERVER : runs
+    OSQUERYINSTANCE ||--|| EXTENSIONCLIENT : runs
+    EXTENSIONSERVER ||--|| EXTENSION : forwards
+    OSQUERYRUNNER ||--|| STATS: updates
+```
+
+The below diagram shows the desired future state -- it is identical to the above diagram except for the relationship cardinalities (e.g. `OSQUERYRUNNER` runs 0 or more of `OSQUERYINSTANCE`, rather than exactly 1 `OSQUERYINSTANCE`).
+
+```mermaid
+---
+title: Future state (0-n osquery instances)
+---
+erDiagram
+    RUNGROUP ||--|| OSQUERYRUNNER : runs
+    OSQUERYRUNNER ||--o{ OSQUERYINSTANCE: runs
+    OSQUERYINSTANCE ||--|| EXTENSION : owns
+    EXTENSION }o--|| JSONRPCCLIENT : uses
+    OSQUERYINSTANCE ||--|{ EXTENSIONSERVER : runs
+    OSQUERYINSTANCE ||--|| EXTENSIONCLIENT : runs
+    EXTENSIONSERVER ||--|| EXTENSION : forwards
+    OSQUERYRUNNER ||--o{ STATS: updates
+```
+
+## Consequences
+
+The osquery instance will now be responsible for managing extension health -- if the extension exits, the instance will need to take corrective action. (This is potentially a better outcome than before, since now extension shutdown can be remediated without a full launcher restart.)
+
+This decision does preserve a shared JSONRPC client to talk to Kolide SaaS. If we determine this is not optimal, it would be fairly easy to update to a model where we have one JSONRPC client per extension.
```

Comment:
any known reasons to share vs break out per extension? or we're just not aware of a reason to duplicate the client

Reply:
> or we're just not aware of a reason to duplicate the client

Basically this! I am imagining it's the sort of thing that I'll have a stronger opinion on once we get further through implementation. But either way, not hard to change after the fact.